workshop/stringBasic/StringBasic.cpp

```diff
@@ -0,0 +1,245 @@
+    if (m_size == 0) {
+        m_size = 2;
+        m_string = new char[m_size];
+        std::memset(m_string, 0, m_size);
+        m_string[m_lenght] = chr;
+        m_lenght++;;
+    }
+    else if (m_lenght == (m_size - 1)) {
+        increase_size(SIZE_SCALE * m_size);
+    }
+    else {
+        m_string[m_lenght] = chr;
+        m_lenght++;
+    }
+}
+
+
+void StringBasic::add_string(const char* const str) {
+    for (std::size_t i = 0; str[i]; i++) {
+        add_char(str[i]);
+    }
+}
+
+
+void StringBasic::insert(StringBasic& str, std::size_t offset) {
+    if (offset > m_lenght) {
+        throw std::out_of_range("invalid offset");
+    }
+    if ((m_lenght + str.m_lenght) > (m_size - 1)) {
+        increase_size(SIZE_SCALE + str.m_lenght);
+    }
+
+    std::memmove(m_string + offset + str.m_lenght, m_string + offset, m_lenght - offset);
+    std::memcpy(m_string + offset, str.m_string, str.m_lenght);
+    m_lenght += str.m_lenght;
+}
+
+
+void StringBasic::replace(StringBasic& substring, StringBasic& newsubstring) {
+    for (std::size_t i = 0; i <= (m_lenght - substring.m_lenght); i++) {
+        bool is_substring = true;
+        for (std::size_t j = 0; j < substring.m_lenght; j++) {
+            if (m_string[i + j] != substring.m_string[j]) {
+                is_substring = false;
+                break;
+            }
+        }
+        if (is_substring) {
+            if ((m_size - 1) < (m_lenght + newsubstring.m_lenght - substring.m_lenght)) {
+                increase_size(m_lenght + newsubstring.m_lenght - substring.m_lenght);
+            }
+
+            if (newsubstring.m_lenght > substring.m_lenght) {
+                std::memmove(m_string + i + newsubstring.m_lenght, m_string + i + substring.m_lenght, m_size - (i + newsubstring.m_lenght + 1));
+                std::memcpy(m_string + i, newsubstring.m_string, newsubstring.m_lenght);
+                m_lenght += (newsubstring.m_lenght - substring.m_lenght);
+                std::memset(m_string + m_lenght, 0, m_size - m_lenght);
+            }
+            else if (newsubstring.m_lenght < substring.m_lenght) {
+                std::memcpy(m_string + i, newsubstring.m_string, newsubstring.m_lenght);
+                std::memmove(m_string + i + newsubstring.m_lenght, m_string + i + newsubstring.m_lenght + 1, m_size - (i + newsubstring.m_lenght + 1));
+                m_lenght -= (substring.m_lenght - newsubstring.m_lenght);
+                std::memset(m_string + m_lenght, 0, m_size - m_lenght);
+            }
+            else {
+                std::memcpy(m_string + i, newsubstring.m_string, newsubstring.m_lenght);
+            }
+            i += newsubstring.m_lenght;
+        }
+    }
+}
+
+
+void StringBasic::increase_size(std::size_t increasment) {
+    m_size += increasment;
+    char* new_string = new char[m_size];
+    std::memset(new_string, 0, m_size);
+    std::memcpy(new_string, m_string, m_lenght);
+
+    try {
```

Comment:
I don't think delete can throw, and even if it can, you should wrap these 4 lines since you're using them in multiple places around this class which bloats the code for no reason.